Add missing_code_block_language_in_doc_comment lint to flutter/packages.

kallentu opened the pull request:
Adds this Dartdoc-related lint to the flutter repository, in replacement of the Dartdoc warning (missingCodeBlockLanguage) because it will be deprecated and removed soon.

flutter/flutter already has this lint as well. Adding to flutter/engine with https://github.com/flutter/engine/pull/51944.

Lint Proposal: https://github.com/dart-lang/linter/issues/4904

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I [linked to at least one issue that this PR fixes] in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

stuartmorgan commented:
Is this lint in 3.4, or do we need to wait another stable release cycle?

Piinks commented:
(PR triage): What is the status of this PR?

kallentu commented:
> (PR triage): What is the status of this PR?

@Piinks 
Linux analyze legacy N-1 is failing because it's linting on the wrong things. There's a commit that just so happens to fall in 3.5 (https://github.com/dart-lang/sdk/commit/b6ac281270ac751175c5298ce78dd9f2cb3bd594) which isn't picked up in that tryjob. The rest of the lint work is done in 3.4 and the version of legacy N-1 is at Dart 3.4.x.

What's the best way forward? I don't want to be doing the fix for what's failing because it's wrong behaviour. I'm happy to wait longer, but I suspect N-1 will become N-2 and that will end up failing.

Piinks commented:
Hmm, I am not sure. @stuartmorgan would likely be best to advise.
I think you would have to bump the dependency on each affected package to make it no longer compatible with N-1 and N-2.

stuartmorgan commented:
The options are:
- Drop support for N-1 in those packages
- Wait for the next stable to do this, and drop N-2 in those packages
- Wait for two more stables (I would not recommend this one; the value of N-2 support is low)
- Add `// ignore` annotations for the N-1 false positives, with a comment explaining why and linking to an issue that tracks removing them after the next stable update. There's no warning for unnecessary `// ignore`s, so that will work on all versions.

Any of those is fine. Our policy is that we only commit to supporting N, so the bar for dropping N-1 is low; N-1/N-2 tests are to prevent *accidentally* doing something that doesn't compile in N-1/N-2 and shipping it in a version that claims to support that version, thus breaking clients.

kallentu commented:
> The options are:
> 
> * Drop support for N-1 in those packages
> * Wait for the next stable to do this, and drop N-2 in those packages
> * Wait for two more stables (I would not recommend this one; the value of N-2 support is low)
> * Add `// ignore` annotations for the N-1 false positives, with a comment explaining why and linking to an issue that tracks removing them after the next stable update. There's no warning for unnecessary `// ignore`s, so that will work on all versions.
> 
> Any of those is fine. Our policy is that we only commit to supporting N, so the bar for dropping N-1 is low; N-1/N-2 tests are to prevent _accidentally_ doing something that doesn't compile in N-1/N-2 and shipping it in a version that claims to support that version, thus breaking clients.

Added ignores as you suggested with the issue tracking it: https://github.com/flutter/flutter/issues/157620.

Fixed up pubspec and changelogs. Should be ready for review and to move forward.

stuartmorgan commented on packages/flutter_migrate/CHANGELOG.md:
```diff
@@ -1,5 +1,6 @@
-## NEXT
+## 0.0.1+4
 
+* Added `missing_code_block_language_in_doc_comment` lint.
```
Nit: Adds

(Per the CHANGELOG style guide in the contributing docs.)

stuartmorgan replied:
Same for the other CHANGELOGs as well.

kallentu replied:
Done, thanks for the reminder!

stuartmorgan commented on packages/process/CHANGELOG.md:
```diff
@@ -1,5 +1,6 @@
-## NEXT
+## 5.0.3
```
The changelog and pubspec changes for this package can be reverted, since now you're only adding a non-doc comment to an implementation file, which doesn't affect clients.

stuartmorgan reviewed (approved):
LGTM with minor changes.

chunhtai reviewed (approved):
LGTM

ditman reviewed (approved):
LGTM, but there's some TODOs that could be done now I think?

kallentu commented on packages/rfw/lib/src/flutter/material_widgets.dart:
```diff
@@ -64,10 +64,16 @@ import 'runtime.dart';
 ///
+// TODO(kallentu): Remove ignore and fix when stable is bumped.
+// https://github.com/flutter/flutter/issues/157620
+// ignore: missing_code_block_language_in_doc_comment
 ///    * `buttonMinWidth`
 ///    * `buttonHeight`
 ///    * `buttonAlignedDropdown`
```
Yeah, this is known. It'll get fixed next release. https://github.com/flutter/flutter/issues/157620 explains a bit, but the commit that fixes it isn't in the right version so the n-1 tryjob fails. Solving this by putting a lint ignore for now. Will follow up later to clean it up :)

kallentu commented on packages/webview_flutter/webview_flutter_web/lib/src/http_request_factory.dart:
```diff
@@ -39,6 +39,9 @@ class HttpRequestFactory {
   ///
+  // TODO(kallentu): Remove ignore and fix when stable is bumped.
+  // https://github.com/flutter/flutter/issues/157620
+  // ignore: missing_code_block_language_in_doc_comment
```
Yeah fair enough. I can do that. I think I was hoping to just preserve the comments as are for now. But no harm, I'll fix these indented comments --> we prefer those anyways,

kallentu commented on packages/rfw/lib/src/dart/binary.dart:
```diff
@@ -159,6 +171,9 @@ Uint8List encodeLibraryBlob(RemoteWidgetLibrary value) {
+// TODO(kallentu): Remove ignore and fix when stable is bumped.
+// https://github.com/flutter/flutter/issues/157620
+// ignore: missing_code_block_language_in_doc_comment
 ///       02 00 00 00 00 00 00 00  01 00 00 00 00 00 00 00
```
Sure. I can make that change.

ditman commented:
Thanks @kallentu!!
